bundle/config/mutator/resourcemutator/resource_mutator.go

```diff
@@ -0,0 +1,315 @@
+		snapshot = root
+
+		return selectResources(root, addedResources)
+	})
+	if err != nil {
+		return diags.Extend(diag.Errorf("failed to select resources: %s", err))
+	}
+
+	diags = diags.Extend(applyNormalizeMutators(ctx, b))
+	if diags.HasError() {
+		return diags
+	}
+
+	diags = diags.Extend(applyInitializeMutators(ctx, b))
+	if diags.HasError() {
+		return diags
+	}
+
+	// after mutators, we merge updated resources back to snapshot to preserve non-selected resources
+	err = b.Config.Mutate(func(root dyn.Value) (dyn.Value, error) {
+		return mergeResources(root, snapshot)
+	})
+	if err != nil {
+		return diags.Extend(diag.Errorf("failed to merge resources: %s", err))
+	}
+
+	return diags
+}
+
+// NormalizeResources normalizes resources specified resources,
+// and should be used by mutators after they have updated resources.
+func NormalizeResources(
+	ctx context.Context,
+	b *bundle.Bundle,
+	updatedResources ResourceKeySet,
+) diag.Diagnostics {
+	if updatedResources.IsEmpty() {
+		return nil
+	}
+
+	var diags diag.Diagnostics
+	var snapshot dyn.Value
+
+	err := b.Config.Mutate(func(root dyn.Value) (dyn.Value, error) {
+		snapshot = root
+
+		return selectResources(root, updatedResources)
+	})
+	if err != nil {
+		return diags.Extend(diag.Errorf("failed to select resources: %s", err))
+	}
+
+	diags = diags.Extend(applyNormalizeMutators(ctx, b))
+	if diags.HasError() {
+		return diags
+	}
+
+	// after mutators, we merge updated resources back to snapshot to preserve non-selected resources
+	err = b.Config.Mutate(func(root dyn.Value) (dyn.Value, error) {
+		return mergeResources(root, snapshot)
+	})
+	if err != nil {
+		return diags.Extend(diag.Errorf("failed to merge resources: %s", err))
+	}
+
+	return diags
+}
+
+// selectResources returns bundle configuration with resources only present in resourcePaths.
+func selectResources(root dyn.Value, resourcePaths ResourceKeySet) (dyn.Value, error) {
+	resourcesKeyString := "resources"
+	resourcesPath := dyn.NewPath(dyn.Key(resourcesKeyString))
+
+	newRoot := root
+	var err error
+
+	// remove resource types that are not in resourcePaths
+	newRoot, err = dyn.MapByPath(
+		newRoot,
+		resourcesPath,
+		func(p dyn.Path, resources dyn.Value) (dyn.Value, error) {
+			return merge.Select(resources, resourcePaths.Types())
+		},
+	)
+	if err != nil {
+		return dyn.InvalidValue, err
+	}
+
+	// for each resource type, remove resources by name
+	for _, resourceType := range resourcePaths.Types() {
+		resourceTypePath := resourcesPath.Append(dyn.Key(resourceType))
+
+		newRoot, err = dyn.MapByPath(
+			newRoot,
+			resourceTypePath,
+			func(p dyn.Path, resource dyn.Value) (dyn.Value, error) {
+				return merge.Select(resource, resourcePaths.Names(resourceType))
+			},
+		)
+		if err != nil {
+			return dyn.InvalidValue, err
+		}
+	}
+
+	return newRoot, err
+}
+
+// mergeResources returns bundle configuration by merging all resources from src into dst,
+// overriding existing resources if they exist.
```

Comment:
Does this function accepts full bundle configuration? What does it do with things outside of 'resources'? 

Should it even accept full bundle? if it's called mergeResources, I'd expect it to be applied only on resources subtrees.

Reply:
We need to preserve parts of the configuration. For example, "presets", "variables", and other properties. We don't have a way to partially mutate the bundle because we use `b.Config.Mutate`.

Reply:
Does not MapByPattern/MapByPath allow you to mutate the bundle partially?

Reply:
No, because Map function selects a simple path, while we need a sub-tree. For example:

Input
```
variables:
  important_variable: ...
resources: 
  jobs:
    job_1: ...
    job_2: ...
  pipelines:
    pipeline_1: ...
    pipeline_2: ...
```

And only job_2 and pipeline_2 were added, we should run mutators using:

```
variables:
  important_variable: ...
resources: 
  jobs:
    job_2: ...
  pipelines:
    pipeline_2: ...
```